libafl_bolts/src/ownedref.rs

```diff
@@ -815,6 +869,213 @@ impl<'a, T> From<&'a mut &'a mut [T]> for OwnedMutSlice<'a, T> {
     }
 }
 
+/// Wrap a mutable slice and convert to a Box on serialize.
+/// We use a hidden inner enum so the public API can be safe,
+/// unless the user uses the unsafe [`OwnedMutSizedSlice::from_raw_mut`].
+/// The variable length version is [`OwnedMutSlice`].
+#[derive(Debug)]
+pub enum OwnedMutSizedSliceInner<'a, T: 'a + Sized, const N: usize> {
+    /// A raw ptr to a memory location of length N
+    RefRaw(*mut [T; N], UnsafeMarker),
+    /// A ptr to a mutable slice of the type
+    Ref(&'a mut [T; N]),
+    /// An owned [`Box`] of the type
+    Owned(Box<[T; N]>),
+}
+
+impl<'a, T: 'a + Sized + Serialize, const N: usize> Serialize
+    for OwnedMutSizedSliceInner<'a, T, N>
+{
+    fn serialize<S>(&self, se: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        match self {
+            OwnedMutSizedSliceInner::RefRaw(rr, _) => unsafe { &**rr }.serialize(se),
+            OwnedMutSizedSliceInner::Ref(r) => (*r).serialize(se),
+            OwnedMutSizedSliceInner::Owned(b) => (*b).serialize(se),
+        }
+    }
+}
+
+impl<'de, 'a, T: 'a + Sized, const N: usize> Deserialize<'de> for OwnedMutSizedSliceInner<'a, T, N>
+where
+    T: Deserialize<'de>,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        arrays::deserialize(deserializer).map(OwnedMutSizedSliceInner::Owned)
+    }
+}
+
+/// Wrap a mutable slice of constant size N and convert to a Box on serialize
+#[allow(clippy::unsafe_derive_deserialize)]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct OwnedMutSizedSlice<'a, T: 'a + Sized, const N: usize> {
+    inner: OwnedMutSizedSliceInner<'a, T, N>,
+}
+
+impl<'it, T, const N: usize> IntoIterator for &'it mut OwnedMutSizedSlice<'_, T, N> {
+    type Item = <IterMut<'it, T> as Iterator>::Item;
+    type IntoIter = IterMut<'it, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.as_sized_slice_mut().iter_mut()
+    }
+}
+
+impl<'it, T, const N: usize> IntoIterator for &'it OwnedMutSizedSlice<'_, T, N> {
+    type Item = <Iter<'it, T> as Iterator>::Item;
+    type IntoIter = Iter<'it, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.as_sized_slice().iter()
+    }
+}
+
+impl<'a, T: 'a + Sized, const N: usize> OwnedMutSizedSlice<'a, T, N> {
+    /// Create a new [`OwnedMutSizedSlice`] from a raw pointer
+    ///
+    /// # Safety
+    ///
+    /// The pointer must be valid and point to a map of the size `size_of<T>() * N`
+    /// The content will be dereferenced in subsequent operations.
+    #[must_use]
+    pub unsafe fn from_raw_mut(ptr: NonNull<T>) -> OwnedMutSizedSlice<'a, T, N> {
+        Self {
+            inner: OwnedMutSizedSliceInner::RefRaw(
+                ptr.as_ptr() as *mut [T; N],
```

Comment:
Wait, are you sure you can do this cast on a `*mut u8`? Does rust guarantee it? Sounds a bit fishy to me..

Reply:
afaik converting a `*mut T` to a `*mut [T;N]` is legal (as long as the underlying ptr has enough space ofc).
i searched back in the time but nothing seems to deny this.
also, i used `as` here, not `transmute`. so i guess if it was really a problem the compiler would complain anyway

Reply:
I asked Gemini and the AI agrees :D
FWIW `as` isn't really checked by the compiler, either.

Reply:
it should be to some extend. for raw pointer casting it's limited but your should be able to spot more problems than with transmute. for example, `as` enforces lifetime where transmute doesn't. I guess there are other things?

Reply:
you can totally cast away lifetimes with `as` :P (You may just need to convert the pointer two times)
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5336b4bffe57d74e2d1b00ce9f853c8e